src/Compilers/CSharp/Portable/Compilation/CSharpCompilation.cs

```diff
@@ -3259,6 +3260,78 @@ internal override StrongNameKeys StrongNameKeys
             return true;
         }
 
+        private class DuplicateFilePathsVisitor : CSharpSymbolVisitor
+        {
+            // note: the default HashSet<string> uses an ordinal comparison
+            private readonly PooledHashSet<string> _duplicatePaths = PooledHashSet<string>.GetInstance();
+
+            private readonly DiagnosticBag _diagnostics;
+
+            private bool _hasDuplicateFilePaths;
+
+            public DuplicateFilePathsVisitor(DiagnosticBag diagnostics)
+            {
+                _diagnostics = diagnostics;
+            }
+
+            public bool CheckDuplicateFilePathsAndFree(ImmutableArray<SyntaxTree> syntaxTrees, NamespaceSymbol globalNamespace)
+            {
+                var paths = PooledHashSet<string>.GetInstance();
+                foreach (var tree in syntaxTrees)
+                {
+                    if (!paths.Add(tree.FilePath))
+                    {
+                        _duplicatePaths.Add(tree.FilePath);
+                    }
+                }
+                paths.Free();
+
+                if (_duplicatePaths.Any())
+                {
+                    VisitNamespace(globalNamespace);
+                }
+                _duplicatePaths.Free();
+                return _hasDuplicateFilePaths;
+            }
+
+            public override void VisitNamespace(NamespaceSymbol symbol)
+            {
+                foreach (var childSymbol in symbol.GetMembers())
+                {
+                    switch (childSymbol)
+                    {
+                        case NamespaceSymbol @namespace:
+                            VisitNamespace(@namespace);
+                            break;
+                        case NamedTypeSymbol namedType:
+                            VisitNamedType(namedType);
+                            break;
+                    }
+                }
+            }
+
+            public override void VisitNamedType(NamedTypeSymbol symbol)
+            {
+                Debug.Assert(symbol.ContainingSymbol.Kind == SymbolKind.Namespace); // avoid unnecessary traversal of nested types
+                if (symbol.AssociatedFileIdentifier is not null)
+                {
+                    var location = symbol.Locations[0];
+                    var filePath = location.SourceTree?.FilePath;
+                    if (_duplicatePaths.Contains(filePath!))
+                    {
+                        _diagnostics.Add(ErrorCode.ERR_FileTypeNonUniquePath, location, symbol, filePath);
```

Comment:
>_diagnostics.Add(ErrorCode.ERR_FileTypeNonUniquePath, location, symbol, filePath); [](http://example.com/codeflow?start=24&length=82)

It looks like we are actually going to complain when there is no real conflict between the types. I.e. the only syntax tree among the duplicates declares local types, others do not. Is this behavior intentional? Do we have a test for a scenario like that?

Reply:
I think it is better to block declaration of the file type when another syntax tree has the same path, even if the type's metadata name would not conflict with any other type, for example. This is because in the current implementation, the binder will permit usage of the file type in any of the trees which have the same path, which might allow unintended access to the file type.

Adding a test to demonstrate this.